sweep/sweeper.go

```diff
@@ -510,6 +510,82 @@ func (s *UtxoSweeper) feeRateForPreference(
 	return feeRate, nil
 }
 
+// removeLastSweepDescendants removes any transactions from the wallet that
+// spend outputs produced by the passed spendingTx. This needs to be done in
+// cases where we're not the only ones that can sweep an output, but there may
+// exists unconfirmed spends that spend outputs created by a sweep transaction.
+// The most common case for this is when someone sweeps our anchor outputs
+// after 16 blocks.
+func (s *UtxoSweeper) removeLastSweepDescendants(spendingTx *wire.MsgTx) error {
+	// Obtain all the past sweeps that we've done so far. We'll need these
+	// to ensure that if the spendingTx spends any of the same inputs, then
+	// we remove any transaction that may be spending those inputs from the
+	// wallet.
+	//
+	// TODO(roasbeef): can be last sweep here if we remove anything confirmed
+	// from the store?
+	pastSweepHashes, err := s.cfg.Store.ListSweeps()
+	if err != nil {
+		return err
+	}
+
+	log.Debugf("Attempting to remove descendant txns invalidated by "+
+		"(txid=%v): %v", spendingTx.TxHash(), spew.Sdump(spendingTx))
+
+	// Construct a map of the inputs this transaction spends for each look
+	// up.
+	inputsSpent := make(map[wire.OutPoint]struct{})
+	for _, txIn := range spendingTx.TxIn {
+		inputsSpent[txIn.PreviousOutPoint] = struct{}{}
+	}
+
+	// We'll now go through each past transaction we published during this
+	// epoch and cross reference the spent inputs. If there're any inputs
+	// in common with the inputs the spendingTx spent, then we'll remove
+	// those.
+	//
+	// TODO(roasbeef): need to start to remove all transaction hashes after
+	// every N blocks (assumed point of no return)
+	for _, sweepHash := range pastSweepHashes {
+		sweepTx, err := s.cfg.Wallet.FetchTx(sweepHash)
+		if err != nil {
+			return err
+		}
+
+		// Transaction wasn't found in the wallet, may have already
+		// been
+		// replaced/removed.
+		if sweepTx == nil {
```

Comment:
Need to get more familiar with how sweeps are handled myself. Meanwhile if a tx is replaced, do we remove it from our wallet db?

Reply:
In certain conditions we do, this was more of a defensive check to make sure that we wouldn't run into a panic. 

Internally in teh wallet, it'll try to [rebroadcast anything that it thinks is still unconfirmed on start up](https://github.com/btcsuite/btcwallet/blob/master/wallet/wallet.go#L2880), but we may remove the transaction right there if we can get a detailed backend error: https://github.com/btcsuite/btcwallet/blob/50ea8d1125561279f9c9af1e46bea16ca414b4a5/wallet/wallet.go#L3669-L3687

Reply:
I suspect this patch might actually fix other issues in the tracker, that are a result of users spending unconfirmed inputs generated by a sweeper sweep.

Reply:
Cool thanks for the detailed explanation!